### PR TITLE
feat(linter/eslint): implement `id-match` rule

### PR DESCRIPTION
**Pull request by v-sayapin**
This PR implements the `eslint/id-match` rule, which enforces a configured naming regex for identifiers

ESLint ref: https://eslint.org/docs/latest/rules/id-match

Issue: https://github.com/oxc-project/oxc/issues/479

### Behavior Differences from ESLint

This implementation is intentionally stricter in two cases:

- Computed destructuring keys are checked, for example `const { [bad_name]: x } = obj` and `({ [bad_name]: x } = obj)`, because the computed key is a normal reference expression, not a binding introduced by destructuring
- With `properties` enabled, ordinary top-level dynamic import option keys are checked, for example `import("x", { bad_option: true })`. Import attributes inside `with { ... }` are still ignored

### TypeScript

TypeScript syntax is supported on a best-effort basis for identifiers that flow through the same visited AST node kinds and transparent wrappers such as `as`, `satisfies`, non-null assertions, and instantiation expressions

This is not a replacement for `@typescript-eslint/naming-convention` or future tsgolint naming-convention work: https://github.com/oxc-project/tsgolint/issues/186

### Regex Compatibility

The pattern is compiled with Rust regex syntax. JavaScript-specific regex features such as lookaround and backreferences are not supported. Unicode escapes should use Rust syntax, for example `\u{XXXX}`

### Algorithmic Complexity

The generated rule runner limits the hot path to identifier-like AST nodes:

- `BindingIdentifier`
- `IdentifierReference`
- `IdentifierName`
- `PrivateIdentifier`
- `LabelIdentifier`

Each visited identifier first runs the regex check and returns immediately when the name matches

For non-default configurations, the hot path is linear in the total length of checked identifier names, with additional ancestor/context checks only for failed regex matches

Space complexity is `O(1)` beyond the compiled regex and options. The hot path does not allocate unless a diagnostic is emitted

The default pattern `^.+$` matches every non-empty identifier name, so `should_run` skips rule work entirely for the default configuration

### AI Disclosure

Claude Opus 4.7 & Codex GPT 5.5

- Code review
- Comparison with ESLint ref rule behavior
- Discussion of edge cases
- Writing and validating tests

### CODE REVIEW

**Comment by v-sayapin** [comment]
@codex review, use the repository skill $performance-lint-rules

File: crates/oxc_linter/src/rules/eslint/id_match.rs [hunk 1]
```diff
@@ -0,0 +1,1385 @@
...
+            Regex::new(&pattern).map_err(|err| serde_json::Error::custom(err.to_string()))?;
+
+        Ok(Self(Box::new(CompiledIdMatchConfig { regex, options })))
+    }
+
+    fn should_run(&self, _ctx: &ContextHost) -> bool {
+        // The default pattern matches every non-empty identifier name, so the
+        // rule cannot report anything. Skip rule work in that case.
+        self.regex.as_str() != DEFAULT_PATTERN
+    }
```
Comment (camc314) [inline crates/oxc_linter/src/rules/eslint/id_match.rs:166]: this doesn't feel right. `regex` should probably just be an option, and we can simplify this to self.regex.is_some

File: crates/oxc_linter/src/rules/eslint/id_match.rs [hunk 2]
```diff
@@ -0,0 +1,1385 @@
...
+
+impl Rule for IdMatch {
+    fn from_configuration(value: Value) -> Result<Self, serde_json::error::Error> {
+        let RawIdMatchConfig(pattern, options) =
+            serde_json::from_value::<TupleRuleConfig<RawIdMatchConfig>>(value)
+                .map(TupleRuleConfig::into_inner)?;
+        let regex =
+            Regex::new(&pattern).map_err(|err| serde_json::Error::custom(err.to_string()))?;
+
+        Ok(Self(Box::new(CompiledIdMatchConfig { regex, options })))
```
Comment (camc314) [inline crates/oxc_linter/src/rules/eslint/id_match.rs]: you should be able to deserialize the regex directly inside the rule config - take a look at some of the other rules for ideas

File: crates/oxc_linter/src/rules/eslint/id_match.rs [hunk 3]
```diff
@@ -0,0 +1,1385 @@
...
+        }
+        AssignmentTargetMaybeDefault::AssignmentTargetWithDefault(target) => {
+            assignment_target_identifier_name_from_target(&target.binding)
+        }
+        _ => None,
+    }
+}
+
+fn assignment_target_identifier_name_from_target<'a>(
+    target: &'a oxc_ast::ast::AssignmentTarget<'a>,
```
Comment (camc314) [inline crates/oxc_linter/src/rules/eslint/id_match.rs]: please import these at the top level rather than specifying the path here

File: crates/oxc_linter/src/rules/eslint/id_match.rs [hunk 4]
```diff
@@ -0,0 +1,1385 @@
...
+fn transparent_reference_parent<'a, 'b>(
+    node: &'b AstNode<'a>,
+    ctx: &'b LintContext<'a>,
+) -> (&'b AstNode<'a>, Span) {
+    let mut current = node;
+    let mut span = node.span();
+
+    loop {
+        let parent = ctx.nodes().parent_node(current.id());
+        match parent.kind() {
```
Comment (camc314) [inline crates/oxc_linter/src/rules/eslint/id_match.rs]: use iter_ancestor or similar? i believe we have a helper for this

File: crates/oxc_linter/src/rules/eslint/id_match.rs (hunk 1 above)
Comment (v-sayapin) [reply crates/oxc_linter/src/rules/eslint/id_match.rs:166]: Fixed it

File: crates/oxc_linter/src/rules/eslint/id_match.rs (hunk 2 above)
Comment (v-sayapin) [reply crates/oxc_linter/src/rules/eslint/id_match.rs]: Thanks!
Moved regex compilation into serde deserialization on the config itself, so `from_configuration` now only deserializes `TupleRuleConfig<IdMatchConfig>`
I used a local deserializer here because the first tuple element is optional only when omitted
`[null]` should be rejected rather than treated as a silent disable, which is closer to ESLint

File: crates/oxc_linter/src/rules/eslint/id_match.rs (hunk 3 above)
Comment (v-sayapin) [reply crates/oxc_linter/src/rules/eslint/id_match.rs]: Done

File: crates/oxc_linter/src/rules/eslint/id_match.rs (hunk 4 above)
Comment (v-sayapin) [reply crates/oxc_linter/src/rules/eslint/id_match.rs]: You're right!
I simplified this by switching `ctx.nodes().ancestors(node.id())`

**Review by camc314 — APPROVED** [review]
thank you!

File: crates/oxc_linter/src/rules/eslint/id_match.rs [hunk 5]
```diff
@@ -0,0 +1,1388 @@
...
+                    if function.r#type == FunctionType::FunctionDeclaration
+            )
+            && !matches!(effective_parent.kind(), AstKind::VariableDeclarator(_))
+        {
+            return;
+        }
+
+        if matches!(effective_parent.kind(), AstKind::CallExpression(_) | AstKind::NewExpression(_))
+        {
+            return;
```
[the remark this answers is not in the document]
Comment (camc314) [reply crates/oxc_linter/src/rules/eslint/id_match.rs:487]: Verified against ESLint 10.3.0. With the default `id-match` config, this should not report. I reverted the added fail case and restored it to the passing behavior.

File: crates/oxc_linter/src/rules/eslint/id_match.rs [hunk 6]
```diff
@@ -0,0 +1,1388 @@
...
+                {
+                    self.report(ctx, node.span(), ident.name.as_str());
+                }
+                true
+            }
+            AstKind::PrivateFieldExpression(member) => {
+                if member.object.span() == subject_span && self.options().properties {
+                    self.report(ctx, node.span(), ident.name.as_str());
+                }
+                true
```
[the remark this answers is not in the document]
Comment (camc314) [reply crates/oxc_linter/src/rules/eslint/id_match.rs:525]: Same here. ESLint does not report this with `properties: false` / default config, so I removed the added fail case and kept the implementation aligned with ESLint.

File: crates/oxc_linter/src/rules/eslint/id_match.rs [hunk 7]
```diff
@@ -0,0 +1,1388 @@
...
+    /// Whether object literal property names, class method names, and assigned
+    /// member names such as `obj.prop = value` are checked.
+    properties: bool,
+    /// Whether to ignore shorthand and aliased bindings introduced by object
+    /// destructuring, such as `foo` in `const { foo } = obj` and `alias` in
+    /// `const { foo: alias } = obj`. This does not suppress computed key
+    /// references such as `const { [key]: value } = obj`.
+    ignore_destructuring: bool,
+    /// Whether to check only declaration names. References, member names,
+    /// labels, and function or arrow parameters are skipped.
```
[the remark this answers is not in the document]
Comment (camc314) [reply crates/oxc_linter/src/rules/eslint/id_match.rs:87]: will push a pr
